cmd/ncproxy/ncproxy.go

```diff
@@ -29,6 +29,114 @@ type grpcService struct{}
 
 var _ ncproxygrpc.NetworkConfigProxyServer = &grpcService{}
 
+func (s *grpcService) AssignVF(ctx context.Context, req *ncproxygrpc.AssignVFRequest) (_ *ncproxygrpc.AssignVFResponse, err error) {
```

Comment:
Curious about the need for both `AssignVF` and `AddNICVirtualFunction` as part of setting up a VF NIC. My naïve assumption was that since the standard NIC flow takes a call to `AddNIC`, the VF flow would similarly take a call to something like `AddNICVF`, additionally passing along things like the PCI ID.

What led to us splitting the VF assignment into two functions, and what additional information needs to be passed? If this is due to needing to pass the information that would normally be in the HNS endpoint, have we considered still utilizing an endpoint as a way to track the NIC information? That seems like it could simplify the flow.

Reply:
I split the functions into `AssignVF` and `AddNICVirtualFunction` just to follow the same kind of flow that was done in the HNS scenario, where first the endpoint is created and later the endpoint is added to the pod as a nic. I can discuss with aznet to see if they're okay with this being in a single call instead if you'd prefer? 

> what additional information needs to be passed compared to the standard NIC flow
overall, we additionally need the device instance ID and VF index. We also don't need a network name or any endpoint policies in this flow. 

> If this is due to needing to pass the information that would normally be in the HNS endpoint, have we considered still utilizing an endpoint as a way to track the NIC information?
As in logically from our end making a fake endpoint? Personally, I think that would just lead to more confusion

Reply:
My view here is that exposing both `AssignVF` and `AddNICVirtualFunction` is effectively a leaky abstraction. For all current network configuration, the model is simply create endpoint -> call `AddNIC`. I think it would be best if we could try to stay close to this model here as well.

A few ideas for how we might do this:
- Combine the two functions into a single function that takes the full set of parameters. This would allow us to have a single entrypoint by the network agent for adding such a NIC.
  - My question on this would be, is there any reason the network agent would want this function broken up? e.g. do we anticipate them needing to do `AssignVF` -> some other work on their end -> `AddNICVirtualFunction`?
- Have the network agent first create an endpoint that caches the information that would have been passed via `AddNICVirtualFunction`. Then we can retrieve the information from the endpoint directly when they call our single add function for the VF. This could be handled under the hood in one of two ways:
  - Create an actual underlying HNS endpoint that stores the info.
  - Because the information does not actually need to be used by HNS for the VF case, instead just have ncproxy cache it and pass it along as part of the later add VF NIC call.

One other slight nit: Is there anything in our code that actually cares if they are passing us a VF? Or from our perspective is it just an arbitrary PCI device? If the latter, then I would lean towards avoiding the "VF" term in the API docs, since these would work just as well if given a PF.

Reply:
@kevpar I get what you're saying about a leaky abstraction, but we also expose so much about how an HNS network is created and its setup. I think it makes sense to have these explicit like this since the vnetagent is the one that is handling the logic to initiate these code flows. I think it makes sense to give them slightly lower level abstractions to do that. 

To point one, I think it's fine to keep the calls separate in case there could be a design in the future where we don't need the latter half of the equation (iow calling `AddNICWithVF`). 
To point two, I don't think it really makes sense to try to overload the concept of an HNS endpoint with information on a vf nic. I think it should be very clear that the flows are different. 

As for the naming including `VF`, I think we can change that. Do you have any suggestions on what would be better?